src/FakeItEasy/Configuration/AssertConfigurationBase.cs

```diff
@@ -0,0 +1,28 @@
+namespace FakeItEasy.Configuration
+{
+    using System;
+    using System.Linq.Expressions;
+
+    /// <inheritdoc />
+    /// <summary>
+    /// Provides extension methods for <see cref="IAssertConfiguration"/>.
+    /// </summary>
+    internal abstract class AssertConfigurationBase : IAssertConfiguration

```

Comment:
Is this class intended to remain after you invert the dependency between `Repeated` and the new API? The fact that the root configuration classes (`RuleBuilder`, `AnyCallConfiguration`, `PropertySetterConfiguration`) inherit from feels awkward, IMO... I think it would be better to duplicate the code in this case.

Reply:
I've no strong feelings about it remaining. I waffled over adding it at all. Happy to push the methods down to the (5) concrete classes.
That raises a question about all the testing, though. The "must be null guarded" tests I added went directly at this base class (a bit of a cheat), so I'm considering duplicating (probably parameterizing) for all the concrete instances. Is okay?

And if that's okay, should I do it to the new Specs I added? Currently they just hit `RuleBuilder`, not `AnyCallConfiguration, `PropertySetterConfiguration`, `PropertySetterAdapter`, or `ReturnValueConfiguration`.

Reply:
In fact, I realize what bothers me is the fact that the base class for all configuration classes is named `AssertConfigurationBase`. It makes it look as if all derived classes have something to do with configuration (which they do, but it's only one of their aspects). If we needed to refactor some other aspect of these classes into a base class, the name would no longer fit, and we can't have multiple base classes.

So maybe just renaming the class to `ConfigurationBase` would be enough.

Reply:
Anyway, this class makes sense now because you're relying on the legacy API to implement the new one, but will it still make sense after you refactor?

Reply:
> That raises a question about all the testing, though. The "must be null guarded" tests I added went directly at this base class (a bit of a cheat), so I'm considering duplicating (probably parameterizing) for all the concrete instances. Is okay?
> 
> And if that's okay, should I do it to the new Specs I added? Currently they just hit RuleBuilder, not AnyCallConfiguration,PropertySetterConfiguration,PropertySetterAdapter, or ReturnValueConfiguration.

Good points.

I just looked at the code coverage, the current `MustHaveHappened` method is covered for `RuleBuilder`, `ReturnValueConfiguration` and `AnyCallConfiguration`, but not for `PropertySetterConfiguration`/`PropertySetterAdapter`. Ideally, the specs should cover all implementations, but I realize it's a lot of work... Maybe it can be postponed for now.